### PR TITLE
Fix upgrade, memory leak, abstract sqlite3 interface

### PR DESCRIPTION
**Pull request by jagerman**
I started this PR to just fix the upgrades, but when investigating I
found some problematic memory leaks as well and ended up adding a bunch
of nice abstraction layers to make the code nicer.

I'd normally separate these into multiple commits, but they ended up
quite interdependent to the point where it isn't really feasible to do
so; but here's the details on the changes:

- add a `DEFAULT "register_height"` for the new "update_height" field so
  that the alter table works
- reorganized the statement preparation and database migration so that
  all statement preparation happens *after* migration; otherwise
  statement preparation fails because some of the statements reference
  columns that aren't created until the migration.
- created a `sql_compiled_statement` class that manages sqlite statement
  pointers; there were several places that we were leaking them (the
  main ones were okay, but we also have some dynamically constructed
  ones for variable size queries).
- added a new abstraction layer around sqlite3 binding and value
  extraction that makes the actual sql interaction code much less
  verbose.
- removed the `nettype` argument from `sql_run_statement` as it wasn't
  being used.  (If something in the future needs it, the compiled
  statement class has a reference to the db object, so it can be
  obtained via `statement.nsdb.network_type()`)
- `bind(statement, 1, someval)` now does what
  `sqlite3_bind_whatever(statement, 1, someval, ...)` did before, except
  that it infers the "whatever" from the type (for everything except
  blobs), and does whatever "..." needs to happen.
- `bind(statement, 1, blob_view{someval.data(), size})` can do the same
  for a blob.  (There is also `bind_blob(statement, 1, someval.data(),
  size)` which does the same thing but is nicer in some contexts).
- `auto x = get<T>(statement, 1)` is the bind counterpart for extracting
  a known type (and similarly calls the right sqlite function based on
  `T`).  There is also a `get(statement, 1, x);` that does the same
  thing, but infers the type from whatever `x` is.  `get_blob` similarly
  gets a blob value, though currently this is only used in the existing
  `sql_copy_blob()` wrapper.
- All the `bind` and `get` forms accept an integer enum class as the
  position argument, so the code can now write just:

    bind(statement, lns_db_setting_column::top_height, value);

  instead of:

    bind(statement, static_cast<int>(lns_db_setting_column::top_height), value);

- bind_all lets you bind all the parameters in one go (and also clears
  existing bindings), letting you write things like:

      bind_all(statement, 123, "my string", my_u64_int);

  which previously would have needed:

      sqlite3_clear_bindings(statement);
      sqlite3_bind_int(statement, 1, 123);
      sqlite3_bind_text(statement, 2, "my string", 10, nullptr, nullptr);
      sqlite3_bind_int64(statement, 3, my_u64_int);

- Also there's a `bind_and_run` which combines this binding with a
  sql_run_statement.
- put all of the new binding code in an anonymous namespace, along with
  several existing `static` functions (this is equivalent, just easier
  than having to write static dozens of times).
- removed a superfluous extra "OR" clause in get_mappings_by_owners.  It
  was producing queries such as:

  ... WHERE o1.address IN (?, ?, ?) OR o2.address IN (?, ?, ?) OR o2.address IN (?, ?, ?)

  also simplified that part of the code down quite a bit with a
  pre-allocated string for the "?, ?, ?" part.
- Simplified some some pointer/size pairs with lokimq::string_view
  (which, internally, is just a pointer size pair already).
- Various methods got de-const'ed in this change.  They probably
  shouldn't have been const before because even though they only access
  data, they still mutate internal state of the stored statements, and
  making them all mutable would be a bit gross.  (They did this before,
  too, but since everything was pointers calling into C code I guess
  the compiler just allowed it).

### CODE REVIEW

File: src/cryptonote_core/loki_name_system.cpp [hunk 1]
```diff
@@ -303,17 +520,36 @@ bool mapping_record::active(cryptonote::network_type nettype, uint64_t blockchai
...
 
-  bool result        = prepare_result == SQLITE_OK;
-  if (!result) MERROR("Can not compile SQL statement:\n" << query << "\nReason: " << sqlite3_errstr(prepare_result));
-  return result;
+  if (prepare_result != SQLITE_OK) {
+    MERROR("Can not compile SQL statement:\n" << query << "\nReason: " << sqlite3_errstr(prepare_result));
+    return false;
+  }
+  if (statement)
+    sqlite3_finalize(statement);
```
Comment (Doy-lee) [inline src/cryptonote_core/loki_name_system.cpp]: finalize on nullptr is no-op

File: src/cryptonote_core/loki_name_system.h [hunk 2]
```diff
@@ -157,33 +200,33 @@ struct name_system_db
...
   bool                        prune_db(uint64_t height);
 
-  owner_record                get_owner_by_key      (generic_owner const &owner) const;
-  owner_record                get_owner_by_id       (int64_t owner_id) const;
-  mapping_record              get_mapping           (mapping_type type, std::string const &name_base64_hash) const;
-  std::vector<mapping_record> get_mappings          (std::vector<uint16_t> const &types, std::string const &name_base64_hash) const;
-  std::vector<mapping_record> get_mappings_by_owner (generic_owner const &key) const;
-  std::vector<mapping_record> get_mappings_by_owners(std::vector<generic_owner> const &keys) const;
-  settings_record             get_settings          () const;
+  owner_record                get_owner_by_key      (generic_owner const &owner);
```
Comment (Doy-lee) [inline src/cryptonote_core/loki_name_system.h]: Getting rid of the const here means the tests need to be adjusted

File: src/cryptonote_core/loki_name_system.cpp [hunk 3]
```diff
@@ -1056,7 +1292,7 @@ CREATE TABLE IF NOT EXISTS "mappings" (
     "txid" BLOB NOT NULL,
     "prev_txid" BLOB NOT NULL,
     "register_height" INTEGER NOT NULL,
-    "update_height" INTEGER NOT NULL,
+    "update_height" INTEGER NOT NULL DEFAULT "register_height",
```
Comment (Doy-lee) [inline src/cryptonote_core/loki_name_system.cpp]: This is a bug originally in the previous PR, in that when you alter a table you append the new column to the table at the end.

So then when you go and query the row and try to index into it using the enum (that specifies the wrong index on migrated DB's), it gives you the wrong value.

Comment (jagerman) [reply src/cryptonote_core/loki_name_system.cpp]: Reordered the columns so new and migrated will match.  (You'll have to dump your intermediate sqlite db if you recreated one from scratch with the previous PR).

File: src/cryptonote_core/loki_name_system.h (hunk 2 above)
Comment (jagerman) [reply src/cryptonote_core/loki_name_system.h]: Updated.

File: src/cryptonote_core/loki_name_system.cpp (hunk 1 above)
Comment (jagerman) [reply src/cryptonote_core/loki_name_system.cpp]: Updated.

**Pull request merged**